u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java

```diff
@@ -192,6 +195,113 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+      inputStream.readFully(keyHandle);
+      int counter = inputStream.readInt();
+      byte[] signature = new byte[inputStream.available()];
+      inputStream.readFully(signature);
+
+      return new TransferAccessResponse(controlByte, transferAccessMessages, keyHandle, counter,
+          signature);
+    } catch (IOException e) {
+      throw new U2FException("Error when parsing raw Transfer Access Response", e);
+    } 
+  }
+
+  /**
+   * Returns an an array of TransferAccessMessages parsed from the raw bytes in the inputStream.
+   * 
+   * @param numberOfTransferAccessMessages
+   * @param inputStream: DataInputStream built by an underlying stream of ByteArrayInputStream() of
+   *        raw bytes. We assume this stream currently points to the beginning of the 
+   *        transferAccessMessages in the stream.
+   * @throws U2FException 
+   */
+  private static TransferAccessMessage[] parseTransferAccessMessageChain(
+      int numberOfTransferAccessMessages, DataInputStream inputStream) throws U2FException {
+    int sequenceNumberSize = 1;
+    int publicKeySize = 65;
+    int appIdSize = 32;
+
+    TransferAccessMessage[] transferAccessMessages =
+        new TransferAccessMessage[numberOfTransferAccessMessages];
+
+    try {
+      if (!inputStream.markSupported()) {
+        throw new U2FException("Stream doesn't support mark and reset.");
+      }
+
+      for (int i = 0; i < numberOfTransferAccessMessages; i++) {
+        inputStream.mark(inputStream.available());
+        inputStream.skipBytes(sequenceNumberSize + publicKeySize + appIdSize);
+
+        X509Certificate newAttestationCertificate = (X509Certificate) CertificateFactory
+            .getInstance("X.509").generateCertificate(inputStream);
+        byte[] newAttestationCertificateBytes;
+        try {
+          newAttestationCertificateBytes = newAttestationCertificate.getEncoded();
+        } catch (CertificateEncodingException e) {
+          throw new U2FException("Error when encoding attestation certificate.", e);
+        }
+
+        int lengthOfAttestationCertificate = newAttestationCertificateBytes.length;
+
+        int lengthOfSignatureUsingPrivateKey = inputStream.readUnsignedByte();
+        inputStream.skipBytes(lengthOfSignatureUsingPrivateKey);
+        int lengthOfSignatureUsingAttestationKey = inputStream.readUnsignedByte();
+        inputStream.reset();
+
+        int sizeOfLengthFieldForSignatureUsingPrivateKey = 1;
+        int sizeOfLengthFieldForSignatureUsingAttestationKey = 1;
+        int lengthOfRawTransferAccessMessage = sequenceNumberSize + publicKeySize + appIdSize
+            + lengthOfAttestationCertificate + sizeOfLengthFieldForSignatureUsingPrivateKey
+            + lengthOfSignatureUsingPrivateKey + sizeOfLengthFieldForSignatureUsingAttestationKey
+            + lengthOfSignatureUsingAttestationKey;
+
+        byte[] rawTransferAccessMessage = new byte[lengthOfRawTransferAccessMessage];
```

Comment:
This is an input stream, not an output stream. I think we may need this to tell it how many bytes to read as it is just reading in the whole thing, not writing it into the buffer in pieces.